File: 백준/Gold/14867. 물통/물통.py
# 물통
import sys
input = sys.stdin.readline
from collections import deque

A, B, C, D = map(int,input().split())

visited = set()

que = deque([])
que.append((0, 0, 0))

while que:
    a, b, cnt = que.popleft()

    if a == C and b == D:
        print(cnt)
        exit()

    if (A, b) not in visited:
        que.append((A, b, cnt+1))
        visited.add((A, b))

    if (a, B) not in visited:
        que.append((a, B, cnt+1))
        visited.add((a, B))

    four_to_b = (max(0, a-(B-b)), min(B, b+a))
    four_to_a = (min(A, a+b), max(0, b - (A-a)))
    
    if four_to_b not in visited:
        que.append((four_to_b[0], four_to_b[1], cnt+1))
        visited.add(four_to_b)

    if four_to_a not in visited:
        que.append((four_to_a[0], four_to_a[1], cnt+1))
        visited.add(four_to_a)

    if (0, b) not in visited:
        que.append((0, b, cnt+1))
        visited.add((0, b))

    if (a, 0) not in visited:
        que.append((a, 0, cnt+1))
        visited.add((a, 0))

print(-1)

File: 백준/Gold/14867. 물통/물통.java
import java.io.*;
import java.util.*;

public class Main {
    static HashSet<Bottle> visited = new HashSet<>();
    static ArrayDeque<Bottle> que = new ArrayDeque<>();

    static class Bottle {
        int a;
        int b;
        int cnt;

        public Bottle(int a, int b, int cnt) {
            this.a = a; this.b = b; this.cnt = cnt;
        }

        @Override
        public int hashCode() {
            return Objects.hash(a, b);
        }

        @Override
        public boolean equals(Object o) {
            Bottle b = (Bottle) o;
            return this.a == b.a && this.b == b.b;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int A = Integer.parseInt(st.nextToken());
        int B = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());
        int D = Integer.parseInt(st.nextToken());

        que.offerLast(new Bottle(0, 0, 0));

        while (!que.isEmpty()) {

            Bottle cur = que.pollFirst();

            if (cur.a == C && cur.b == D) {
                System.out.println(cur.cnt);
                System.exit(0);
            }

            addQue(A, cur.b, cur.cnt+1);
            addQue(cur.a, B, cur.cnt+1);

            addQue(Math.min(A, cur.a + cur.b), Math.max(0, cur.b - (A - cur.a)), cur.cnt+1);
            addQue(Math.max(0, cur.a - (B - cur.b)), Math.min(B, cur.b + cur.a), cur.cnt+1);

            addQue(0, cur.b, cur.cnt+1);
            addQue(cur.a, 0, cur.cnt+1);

        }

        System.out.println(-1);
    }

    public static void addQue(int a, int b, int cnt) {
        Bottle bottle = new Bottle(a, b, cnt);

        if (!visited.contains(bottle)) {
            visited.add(bottle);
            que.offerLast(bottle);
        }
    }
}
